coldata: change Bytes vector implementation

yuzefovich opened the pull request:
**coldata: change Bytes vector implementation**

This commit changes the way `Bytes` vector is implemented in order to
support `Set` operation at arbitrary positions. Now, a `Bytes` vector is
a slice of `element`s.

`element` describes a single `[]byte` value. If a value doesn't exceed
30 bytes, then the whole value is stored within the `element`; otherwise,
the `element` "points" at a region in a common `Bytes.buffer` where the
non-inlined value is stored.

If the value is inlined, then the layout of `element` is used as follows:
```
           24-byte header | 6-byte padding
  element: .............................. | length | true
  Bytes.buffer: N/A
```
where 30 dots describe the inlinable space followed by a single byte for the
length followed by a boolean `true` indicating an inlined value.

If the value is non-inlined, then the layout of `element` is used as follows:
```
                                            padding
  element: .offset. | ..len... | ..cap... | xxxxxx | x | false
  Bytes.buffer: xxxxxxxx | offset .... | xxxxxxxx
```
where first 24 bytes contain our custom "header" of a byte slice that is
backed by `Bytes.buffer`. The following 7 bytes (the padding and the
`inlinedLength` byte) are not used, and then we have a boolean `false`
indicating a non-inlined value.

We have a custom "slice header" for the non-inlined values and don't
just use `[]byte` for two reasons:
1. we can easily overwrite the whole struct when a value is inlined without
causing any issues with the GC (because our slice header doesn't contain an
actual slice, GC doesn't have to track whether the backing array is live or
not);
2. our slice header remains valid even when `Bytes.buffer` is reallocated.
If we were to use `[]byte`, then whenever that reallocation occurred, we
would have to iterate over all non-inlined values and update the byte
slices to point to the new `Bytes.buffer`.

This new implementation was inspired by Velox library from Facebook
(https://facebookincubator.github.io/velox/develop/vectors.html#flat-vectors-scalar-types).

Our original implementation was based on the Arrow format, so we were
getting extremely cheap serialization/deserialization. With this change
some of the benchmarks have regressed significantly:
```
Append/bytes/AppendSimple/NullProbability=0.0-24    356MB/s ± 5%   221MB/s ± 1%    -37.81%  (p=0.000 n=9+8)
Append/bytes/AppendWithSel/NullProbability=0.0-24   241MB/s ± 2%   105MB/s ± 4%    -56.49%  (p=0.000 n=9+10)
Append/bytes/AppendSimple/NullProbability=0.2-24    354MB/s ± 6%   219MB/s ± 4%    -38.11%  (p=0.000 n=8+9)
Append/bytes/AppendWithSel/NullProbability=0.2-24   229MB/s ± 3%    98MB/s ± 4%    -57.20%  (p=0.000 n=9+10)
Copy/bytes/CopySimple/NullProbability=0.0-24       39.7GB/s ± 1%   0.5GB/s ± 0%    -98.67%  (p=0.000 n=10+10)
Copy/bytes/CopyWithSel/NullProbability=0.0-24       753MB/s ± 1%   544MB/s ± 0%    -27.83%  (p=0.000 n=10+8)
Copy/bytes/CopySimple/NullProbability=0.2-24       28.9GB/s ± 1%   0.5GB/s ± 0%    -98.19%  (p=0.000 n=8+10)
Copy/bytes/CopyWithSel/NullProbability=0.2-24       634MB/s ± 0%   575MB/s ± 0%     -9.37%  (p=0.000 n=9+9)
```
```
ArrowBatchConverter/bytes/nullFraction=0.00/BatchToArrow-24   248GB/s ± 2%      3GB/s ± 3%     -98.90%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.25/BatchToArrow-24   240GB/s ± 2%      3GB/s ± 2%     -98.87%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.50/BatchToArrow-24   242GB/s ± 1%      3GB/s ± 2%     -98.86%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.00/ArrowToBatch-24   830GB/s ± 2%      3GB/s ± 2%     -99.59%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.25/ArrowToBatch-24   893GB/s ± 1%      3GB/s ± 1%     -99.62%  (p=0.000 n=10+8)
ArrowBatchConverter/bytes/nullFraction=0.50/ArrowToBatch-24   910GB/s ± 1%      3GB/s ± 1%     -99.62%  (p=0.000 n=10+9)
```
I'm not particularly worried about the regression in the serialization
(since the absolute speeds are still very high for this to not be
a bottle neck), but the slowdown in `Append/Copy` benchmark is
worrisome. I do have some ideas of how to improve those operations to
reduce the regressions to acceptable levels, and I'll add those
optimizations in a follow-up commit.

As a proxy for change of performance of `Bytes` vector overall, I ran
`BenchmarkSortUUID` with results [here](https://gist.github.com/yuzefovich/baaf6b4096cb3e2f1504ad2b38431cc6)
and the unordered distinct benchmarks with results [here](https://gist.github.com/yuzefovich/d8b9166977ec56780cb1749e848f5935).
They show that the number of allocations has decreased in all cases
although the total allocated space increased in some cases with small
number of rows. As a result, there are some regressions in performance
on a small number of rows. Overall, results seem satisfactory.

Notably, as a result of this change the performance of top K sort has
improved drastically when bytes-like types are present in the input
columns (because `CopySlice` used in top K sorter no longer has to move
the data around).

It's worth calling out that we still use the same Arrow format for
serialization, so there is no need to bump the DistSQL version because
the change of this commit is purely to the in-memory representation.

An additional nice result of this change is that we no longer have to
maintain the non-decreasing offsets, which will allow us to remove calls
to `SetLength` in many places where the length of the batch is not
changed. This will be done in the follow-up commit.

Fixes: #75536.

Release note: None

Release justification: low risk, high benefit changes to existing
functionality.

**coldata: optimize Append and Copy for Bytes vectors**

This commit optimizes `Append/Copy` operations for `Bytes` vectors. The
main idea is that a new primitive `copy` is introduced which allows for
copying one `element` from one vector into another `element` of another
vector which removes the allocations when copying inlined values.

An additional improvement is preallocating enough `element`s to
copy/append into many values.
```
name                                               old speed      new speed       delta
Append/bytes/AppendSimple/NullProbability=0.0-24    226MB/s ± 7%    221MB/s ± 8%      ~     (p=0.143 n=10+10)
Append/bytes/AppendWithSel/NullProbability=0.0-24   104MB/s ± 5%    189MB/s ± 1%   +81.57%  (p=0.000 n=10+8)
Append/bytes/AppendSimple/NullProbability=0.2-24    220MB/s ± 2%    229MB/s ± 7%    +4.23%  (p=0.004 n=8+10)
Append/bytes/AppendWithSel/NullProbability=0.2-24  98.4MB/s ± 5%  197.7MB/s ± 4%  +100.98%  (p=0.000 n=10+10)
Copy/bytes/CopySimple/NullProbability=0.0-24        521MB/s ± 0%   2981MB/s ± 1%  +471.63%  (p=0.000 n=10+10)
Copy/bytes/CopyWithSel/NullProbability=0.0-24       539MB/s ± 0%   1020MB/s ± 0%   +89.32%  (p=0.000 n=10+10)
Copy/bytes/CopySimple/NullProbability=0.2-24        518MB/s ± 0%   2913MB/s ± 0%  +462.62%  (p=0.000 n=10+8)
Copy/bytes/CopyWithSel/NullProbability=0.2-24       560MB/s ± 0%    875MB/s ± 0%   +56.12%  (p=0.000 n=10+8)

name                                               old alloc/op   new alloc/op    delta
Append/bytes/AppendSimple/NullProbability=0.0-24     76.1kB ± 8%    118.9kB ± 7%   +56.23%  (p=0.000 n=10+10)
Append/bytes/AppendWithSel/NullProbability=0.0-24     192kB ± 2%      131kB ± 1%   -32.15%  (p=0.000 n=10+8)
Append/bytes/AppendSimple/NullProbability=0.2-24     78.0kB ± 0%    117.2kB ± 3%   +50.27%  (p=0.000 n=8+9)
Append/bytes/AppendWithSel/NullProbability=0.2-24     198kB ± 1%      105kB ± 3%   -47.21%  (p=0.000 n=10+9)
Copy/bytes/CopySimple/NullProbability=0.0-24          0.00B           0.00B           ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.0-24         0.00B           0.00B           ~     (all equal)
Copy/bytes/CopySimple/NullProbability=0.2-24          0.00B           0.00B           ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.2-24         0.00B           0.00B           ~     (all equal)
```
I'm not sure what's up with some of the increase in allocations since
I don't see them locally on mac.

Release note: None

Release justification: low risk, high benefit changes to existing
functionality.

**colexec: cleanup top K with bytes-like types**

Since the previous commit made it so that `Bytes` vectors support Sets
at arbitrary positions, we no longer have to use `CopySlice` in the top
K sort implementation.

Release note: None

Release justification: low risk, high benefit changes to existing
functionality.

**coldata: prohibit Append from the vector into itself**

Previously, `Append`ing from the vector into itself was allowed when
`srcStartIdx == srcEndIdx` in order to be able to truncate the vector.
One of the previous commits removed the only place where this was
actually used (when implementing `Append` for Bytes-like types when
selection vector is present), so this commit removes this no-longer-used
feature.

Release note: None

Release justification: low risk cleanup.

**colexec: remove no longer needed calls to SetLength**

Previously, we required that operators that might populate a `Bytes`
vector set the length on the output batch (which was needed in order to
maintain the invariant of non-decreasing offsets in the old
implementation of the `Bytes` vector) even if the length of the batch
didn't change. This is no longer needed given the updated
implementation, so this commit removes those redundant calls to
`SetLength`.

Release note: None

Release justification: low risk cleanup.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76202)
<!-- Reviewable:end -->

yuzefovich committed:
coldata: change Bytes vector implementation

This commit changes the way `Bytes` vector is implemented in order to
support `Set` operation at arbitrary positions. Now, a `Bytes` vector is
a slice of `element`s.

`element` describes a single `[]byte` value. If a value doesn't exceed
30 bytes, then the whole value is stored within the `element`; otherwise,
the `element` "points" at a region in a common `Bytes.buffer` where the
non-inlined value is stored.

If the value is inlined, then the layout of `element` is used as follows:
```
           24-byte header | 6-byte padding
  element: .............................. | length | true
  Bytes.buffer: N/A
```
where 30 dots describe the inlinable space followed by a single byte for the
length followed by a boolean `true` indicating an inlined value.

If the value is non-inlined, then the layout of `element` is used as follows:
```
                                            padding
  element: .offset. | ..len... | ..cap... | xxxxxx | x | false
  Bytes.buffer: xxxxxxxx | offset .... | xxxxxxxx
```
where first 24 bytes contain our custom "header" of a byte slice that is
backed by `Bytes.buffer`. The following 7 bytes (the padding and the
`inlinedLength` byte) are not used, and then we have a boolean `false`
indicating a non-inlined value.

We have a custom "slice header" for the non-inlined values and don't
just use `[]byte` for two reasons:
1. we can easily overwrite the whole struct when a value is inlined without
causing any issues with the GC (because our slice header doesn't contain an
actual slice, GC doesn't have to track whether the backing array is live or
not);
2. our slice header remains valid even when `Bytes.buffer` is reallocated.
If we were to use `[]byte`, then whenever that reallocation occurred, we
would have to iterate over all non-inlined values and update the byte
slices to point to the new `Bytes.buffer`.

This new implementation was inspired by Velox library from Facebook
(https://facebookincubator.github.io/velox/develop/vectors.html#flat-vectors-scalar-types).

Our original implementation was based on the Arrow format, so we were
getting extremely cheap serialization/deserialization. With this change
some of the benchmarks have regressed significantly:
```
Append/bytes/AppendSimple/NullProbability=0.0-24    356MB/s ± 5%   221MB/s ± 1%    -37.81%  (p=0.000 n=9+8)
Append/bytes/AppendWithSel/NullProbability=0.0-24   241MB/s ± 2%   105MB/s ± 4%    -56.49%  (p=0.000 n=9+10)
Append/bytes/AppendSimple/NullProbability=0.2-24    354MB/s ± 6%   219MB/s ± 4%    -38.11%  (p=0.000 n=8+9)
Append/bytes/AppendWithSel/NullProbability=0.2-24   229MB/s ± 3%    98MB/s ± 4%    -57.20%  (p=0.000 n=9+10)
Copy/bytes/CopySimple/NullProbability=0.0-24       39.7GB/s ± 1%   0.5GB/s ± 0%    -98.67%  (p=0.000 n=10+10)
Copy/bytes/CopyWithSel/NullProbability=0.0-24       753MB/s ± 1%   544MB/s ± 0%    -27.83%  (p=0.000 n=10+8)
Copy/bytes/CopySimple/NullProbability=0.2-24       28.9GB/s ± 1%   0.5GB/s ± 0%    -98.19%  (p=0.000 n=8+10)
Copy/bytes/CopyWithSel/NullProbability=0.2-24       634MB/s ± 0%   575MB/s ± 0%     -9.37%  (p=0.000 n=9+9)
```
```
ArrowBatchConverter/bytes/nullFraction=0.00/BatchToArrow-24   248GB/s ± 2%      3GB/s ± 3%     -98.90%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.25/BatchToArrow-24   240GB/s ± 2%      3GB/s ± 2%     -98.87%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.50/BatchToArrow-24   242GB/s ± 1%      3GB/s ± 2%     -98.86%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.00/ArrowToBatch-24   830GB/s ± 2%      3GB/s ± 2%     -99.59%  (p=0.000 n=10+10)
ArrowBatchConverter/bytes/nullFraction=0.25/ArrowToBatch-24   893GB/s ± 1%      3GB/s ± 1%     -99.62%  (p=0.000 n=10+8)
ArrowBatchConverter/bytes/nullFraction=0.50/ArrowToBatch-24   910GB/s ± 1%      3GB/s ± 1%     -99.62%  (p=0.000 n=10+9)
```
I'm not particularly worried about the regression in the serialization
(since the absolute speeds are still very high for this to not be
a bottle neck), but the slowdown in `Append/Copy` benchmark is
worrisome. I do have some ideas of how to improve those operations to
reduce the regressions to acceptable levels, and I'll add those
optimizations in a follow-up commit.

As a proxy for change of performance of `Bytes` vector overall, I ran
`BenchmarkSortUUID` with results [here](https://gist.github.com/yuzefovich/baaf6b4096cb3e2f1504ad2b38431cc6)
and the unordered distinct benchmarks with results [here](https://gist.github.com/yuzefovich/d8b9166977ec56780cb1749e848f5935).
They show that the number of allocations has decreased in all cases
although the total allocated space increased in some cases with small
number of rows. As a result, there are some regressions in performance
on a small number of rows. Overall, results seem satisfactory.

Notably, as a result of this change the performance of top K sort has
improved drastically when bytes-like types are present in the input
columns (because `CopySlice` used in top K sorter no longer has to move
the data around).

It's worth calling out that we still use the same Arrow format for
serialization, so there is no need to bump the DistSQL version because
the change of this commit is purely to the in-memory representation.

An additional nice result of this change is that we no longer have to
maintain the non-decreasing offsets, which will allow us to remove calls
to `SetLength` in many places where the length of the batch is not
changed. This will be done in the follow-up commit.

Release note: None

Release justification: low risk, high benefit changes to existing
functionality.

jordanlewis commented:
Neat, what is the significant of 19 bytes?

yuzefovich reviewed (commented):
I added the rationale in the commit message (not yet to the PR description), but my thinking is that 16 bytes is the minimum acceptable (so that UUIDs are inlined), and we have at least 3 bytes of padding which gives us 19 bytes. We could add more padding in order to increase inlinable space at the cost of larger allocations, but at the moment a single element is 32 bytes which seems just right to me. It's a bit unfortunate that we cannot use 8 bytes from the slice header, but reusing `SliceHeader.Data` field leads to issues with the GC.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MvPUFRiAqzWWCROvuBM:b-2v4zps)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: cleanup top K with bytes-like types

Since the previous commit made it so that `Bytes` vectors support Sets
at arbitrary positions, we no longer have to use `CopySlice` in the top
K sort implementation.

Release note: None

Release justification: low risk, high benefit changes to existing
functionality.

yuzefovich committed:
colexec: remove no longer needed calls to SetLength

Previously, we required that operators that might populate a `Bytes`
vector set the length on the output batch (which was needed in order to
maintain the invariant of non-decreasing offsets in the old
implementation of the `Bytes` vector) even if the length of the batch
didn't change. This is no longer needed given the updated
implementation, so this commit removes those redundant calls to
`SetLength`.

Release note: None

Release justification: low risk cleanup.

yuzefovich commented:
cc @nvanbenschoten in case you're interested in taking a look at the first commit

yuzefovich committed:
coldata: optimize Append and Copy for Bytes vectors

This commit optimizes `Append/Copy` operations for `Bytes` vectors. The
main idea is that a new primitive `copy` is introduced which allows for
copying one `element` from one vector into another `element` of another
vector which removes the allocations when copying inlined values.

An additional improvement is preallocating enough `element`s to
copy/append into many values.
```
name                                               old speed      new speed       delta
Append/bytes/AppendSimple/NullProbability=0.0-24    226MB/s ± 7%    221MB/s ± 8%      ~     (p=0.143 n=10+10)
Append/bytes/AppendWithSel/NullProbability=0.0-24   104MB/s ± 5%    189MB/s ± 1%   +81.57%  (p=0.000 n=10+8)
Append/bytes/AppendSimple/NullProbability=0.2-24    220MB/s ± 2%    229MB/s ± 7%    +4.23%  (p=0.004 n=8+10)
Append/bytes/AppendWithSel/NullProbability=0.2-24  98.4MB/s ± 5%  197.7MB/s ± 4%  +100.98%  (p=0.000 n=10+10)
Copy/bytes/CopySimple/NullProbability=0.0-24        521MB/s ± 0%   2981MB/s ± 1%  +471.63%  (p=0.000 n=10+10)
Copy/bytes/CopyWithSel/NullProbability=0.0-24       539MB/s ± 0%   1020MB/s ± 0%   +89.32%  (p=0.000 n=10+10)
Copy/bytes/CopySimple/NullProbability=0.2-24        518MB/s ± 0%   2913MB/s ± 0%  +462.62%  (p=0.000 n=10+8)
Copy/bytes/CopyWithSel/NullProbability=0.2-24       560MB/s ± 0%    875MB/s ± 0%   +56.12%  (p=0.000 n=10+8)

name                                               old alloc/op   new alloc/op    delta
Append/bytes/AppendSimple/NullProbability=0.0-24     76.1kB ± 8%    118.9kB ± 7%   +56.23%  (p=0.000 n=10+10)
Append/bytes/AppendWithSel/NullProbability=0.0-24     192kB ± 2%      131kB ± 1%   -32.15%  (p=0.000 n=10+8)
Append/bytes/AppendSimple/NullProbability=0.2-24     78.0kB ± 0%    117.2kB ± 3%   +50.27%  (p=0.000 n=8+9)
Append/bytes/AppendWithSel/NullProbability=0.2-24     198kB ± 1%      105kB ± 3%   -47.21%  (p=0.000 n=10+9)
Copy/bytes/CopySimple/NullProbability=0.0-24          0.00B           0.00B           ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.0-24         0.00B           0.00B           ~     (all equal)
Copy/bytes/CopySimple/NullProbability=0.2-24          0.00B           0.00B           ~     (all equal)
Copy/bytes/CopyWithSel/NullProbability=0.2-24         0.00B           0.00B           ~     (all equal)
```
I'm not sure what's up with some of the increase in allocations since
I don't see them locally on mac.

Release note: None

Release justification: low risk, high benefit changes to existing
functionality.

yuzefovich commented:
I think we can basically ignore the regression in speed of serialization/deserialization since it's still in GB/s range, and I have some improvements for `Append/Copy` (the regressions on which are more worrisome), the numbers are [here](https://gist.github.com/yuzefovich/9deacc920e381ee17ebfcf055f29209f). I'll need more time to polish those up, and I think it's better to not include them into this PR for ease of reviewing anyway.

michae2 reviewed (changes requested):
Spicy! :hot_pepper: :slightly_smiling_face:

I buy your argument about the `ArrowBatchConverter` regressions. Being able to `Bytes.Set()` seems like a big win.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MvQtsMhASqKGbL3rlYp:bl86nfh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/col/coldata/bytes.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQt7z2EFSehoRpEQcX:-MvQsOAu68VCVQ90XcaR:bk57yv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L61)):*
> ```Go
> 	// element was valid (namely, it'll contain a non-inlined value with an
> 	// empty []byte).
> 	nonInlined     bool
> ```

You might be able to inline this, too. (Isn't `data.Data` always nil when the value is inlined and non-nil when the value is pointed to?)
___
*[pkg/col/coldata/bytes.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQtQPWAzr3qU5gZGsK:-MvQtQPWAzr3qU5gZGsL:bekyrml) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L62)):*
> ```Go
> 	// empty []byte).
> 	nonInlined     bool
> 	lengthOrOffset int32
> ```

If this were only used by inlined data, you could also inline it. (E.g. you could use the first inline byte for the length of inlined data.)
___
*[pkg/col/coldata/bytes.go, line 164 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQnobQ6XDcpKnieyBr:-MvQnobQ6XDcpKnieyBs:bt4pmjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L164)):*
> ```Go
> 					e.data = b.buffer[e.lengthOrOffset : e.lengthOrOffset+int32(len(e.data)) : e.lengthOrOffset+int32(cap(e.data))]
> 				}
> 			}
> ```

If `element.data` were something like `{Offset, Len, Cap}` rather than `{Ptr, Len, Cap}` then you wouldn't need to do this update. And bonus, you might be able to use more of `element` for inlining. (This takes go GC out of the loop, of course, which could cause problems if we're relying on go GC to know when to free `Bytes.buffer` based on some `element` outliving its parent `Bytes`. But if an `element` never outlives its parent `Bytes` then it could be fine.)
___
*[pkg/col/coldata/bytes.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQpndTACRbV3N3M7uY:-MvQpndU39rpua94bjgM:b9gvzb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L175)):*
> ```Go
> type Bytes struct {
> 	// elements contains all values stored in this Bytes vector.
> 	elements []element
> ```

I see, you already implemented the "second layer" @msirek was talking about today with this `elements` slice. I wonder how much of the performance changes are due to being able to `Bytes.Set()` using this second layer, rather than due to inlining? The inlining seems like a small optimization on top of this more fundamental change... but maybe I'm misjudging it?

Just curious, did you happen to run any benchmarks with only the second layer, no inlining? If not, would you be willing to try some benchmarks with `BytesMaxInlineLength = -1`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the feedback! I like your suggestions, currently have a WIP to implement them, but will need to polish the code more. I'll put this PR aside for a couple of days and will likely pick it back up on Friday.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MvRtYfT7s1LRBtadMxB:b-rqdp9b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I think it is RFAL. I added a separate commit to add some optimizations (kept separately for ease of reviewing). The performance regressions for `Append/Copy` have been reduced to an acceptable level (and there are some speedups too), here are the overall change of this PR on [SortUUID](https://gist.github.com/yuzefovich/a39850165ed5704b41d684847c816909) and [unordered distinct](https://gist.github.com/yuzefovich/a63766b109a284a1c4f599fc2a72861a).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MvgdTIZ2AWtVWAvO57I:b-mj9yu0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @michae2, and @msirek)

___
*[pkg/col/coldata/bytes.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQt7z2EFSehoRpEQcX:-Mvg-7681dfu3wCUSGDB:byg433c) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L61)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

You might be able to inline this, too. (Isn't `data.Data` always nil when the value is inlined and non-nil when the value is pointed to?)
</blockquote></details>

Refactored.
___
*[pkg/col/coldata/bytes.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQtQPWAzr3qU5gZGsK:-Mvg-8boEK1ksjTPUx9f:b-qz0jif) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L62)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If this were only used by inlined data, you could also inline it. (E.g. you could use the first inline byte for the length of inlined data.)
</blockquote></details>

I decided to keep a single byte for `inlinedLength` separately to reduce the complexity.
___
*[pkg/col/coldata/bytes.go, line 164 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQnobQ6XDcpKnieyBr:-Mvg-Gf63oQPBTygV3Pq:b-guca9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L164)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If `element.data` were something like `{Offset, Len, Cap}` rather than `{Ptr, Len, Cap}` then you wouldn't need to do this update. And bonus, you might be able to use more of `element` for inlining. (This takes go GC out of the loop, of course, which could cause problems if we're relying on go GC to know when to free `Bytes.buffer` based on some `element` outliving its parent `Bytes`. But if an `element` never outlives its parent `Bytes` then it could be fine.)
</blockquote></details>

Nice! This is a very good point, and I implemented your suggestion. Now we have 30 bytes of inlinable space and don't have to update pointers for non-inlined elements when `Bytes.buffer` is reallocated.

We could have squeezed one more byte of inlinable space by merging `inlinedLength` and `nonInlined` boolean into a single `byte`, but I feel like that extra complexity is not worth it. Thoughts?
___
*[pkg/col/coldata/bytes.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQpndTACRbV3N3M7uY:-MvgdHZG215xhS7eavmS:b5sx5fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L175)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I see, you already implemented the "second layer" @msirek was talking about today with this `elements` slice. I wonder how much of the performance changes are due to being able to `Bytes.Set()` using this second layer, rather than due to inlining? The inlining seems like a small optimization on top of this more fundamental change... but maybe I'm misjudging it?

Just curious, did you happen to run any benchmarks with only the second layer, no inlining? If not, would you be willing to try some benchmarks with `BytesMaxInlineLength = -1`?
</blockquote></details>

My guess is that inlining is quite important. I'll run some benchmarks with inlining disabled next week.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @michae2, and @msirek)

___
*[pkg/col/coldata/bytes.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQpndTACRbV3N3M7uY:-Mvt8NFXBtSz9b9d7WpE:b-wx4mqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L175)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

My guess is that inlining is quite important. I'll run some benchmarks with inlining disabled next week.
</blockquote></details>

Hm, the benchmarks [here](https://gist.github.com/yuzefovich/112bedc05a3cb0976a0a807e8c75091b) show that there are some improvements and some regressions without any inlining when comparing against the current tip of this PR.

I think maybe the best approach will be with no inlining, but with preallocating `buffer` with some fixed width (say 16 bytes for each element) right away. I'll prototype that at some point, so hold off on reviewing for now.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
@michae2 PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MxRsprG68gSDRUg-s2d:bxxddyp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @michae2, and @msirek)

___
*[pkg/col/coldata/bytes.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQpndTACRbV3N3M7uY:-MxRsPKH46YMWownh2ON:bqx9qsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L175)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, the benchmarks [here](https://gist.github.com/yuzefovich/112bedc05a3cb0976a0a807e8c75091b) show that there are some improvements and some regressions without any inlining when comparing against the current tip of this PR.

I think maybe the best approach will be with no inlining, but with preallocating `buffer` with some fixed width (say 16 bytes for each element) right away. I'll prototype that at some point, so hold off on reviewing for now.
</blockquote></details>

My initial WIP had a bug where in the `AppendSlice` method which inflated the benchmark numbers. The updated numbers are [here](https://gist.github.com/yuzefovich/112bedc05a3cb0976a0a807e8c75091b), and they are aligned with my expectation - the inlining is important, so I think the PR in its current form is the way to go.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
coldata: prohibit Append from the vector into itself

Previously, `Append`ing from the vector into itself was allowed when
`srcStartIdx == srcEndIdx` in order to be able to truncate the vector.
One of the previous commits removed the only place where this was
actually used (when implementing `Append` for Bytes-like types when
selection vector is present), so this commit removes this no-longer-used
feature.

Release note: None

Release justification: low risk cleanup.

yuzefovich reviewed (commented):
Alright, done pushing, RFAL. The current benchmark numbers are [here](https://gist.github.com/yuzefovich/e62d1f591512b2da16b83b1fed0f3f28).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MxSJn68CAyf5h-ymFgy:b-c6hqki)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @michae2, and @msirek)



<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
This is looking good. I'm still reading but have to pause for a bit.

> Reviewed 6 of 15 files at r1, 1 of 17 files at r5, 5 of 8 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MxgE_LT7lZu64A8htyu:bddr233)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @michae2, and @yuzefovich)

___
*[pkg/col/coldata/bytes.go, line 164 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQnobQ6XDcpKnieyBr:-MxfTu0z720Ovp4vxu84:b5mtwk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L164)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Nice! This is a very good point, and I implemented your suggestion. Now we have 30 bytes of inlinable space and don't have to update pointers for non-inlined elements when `Bytes.buffer` is reallocated.

We could have squeezed one more byte of inlinable space by merging `inlinedLength` and `nonInlined` boolean into a single `byte`, but I feel like that extra complexity is not worth it. Thoughts?
</blockquote></details>

This is fine. No need to squeeze any more byte juice.

nit: But it might be nice to reverse the meaning of the `inlined` boolean so that the zero `element` was a zero-length inlined `[]byte` rather than a zero-length `[]byte` at offset 0 of the buffer. (Maybe you meant to do that, since you wrote `nonInlined` instead of `inlined` in your comment above?)
___
*[pkg/col/coldata/bytes.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQpndTACRbV3N3M7uY:-MxfUb_LDvJXHFUH451i:brxp9sd) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L175)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

My initial WIP had a bug where in the `AppendSlice` method which inflated the benchmark numbers. The updated numbers are [here](https://gist.github.com/yuzefovich/112bedc05a3cb0976a0a807e8c75091b), and they are aligned with my expectation - the inlining is important, so I think the PR in its current form is the way to go.
</blockquote></details>

Sounds good!
___
*[pkg/col/coldata/bytes.go, line 86 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxg9ebE7j83wCKKWnKW:-Mxg9ebF0e4uHUlXYbfd:b-fvv0xu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L86)):*
> ```Go
> // BytesMaxInlineLength is the maximum length of a []byte that can be inlined
> // within element.
> const BytesMaxInlineLength = int(ElementSize - 2)
> ```

nit: Might be a little "safer" to use `unsafe.Offsetof(element.inlinedLength)` instead of `ElementSize - 2` in case go ever puts padding after or between the last two fields.
___
*[pkg/col/coldata/bytes.go, line 186 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxfw3DRCYOJa_5YuD0g:-Mxfw3DRCYOJa_5YuD0h:b6ve6rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L186)):*
> ```Go
> // TODO(yuzefovich): consider renaming it to Slice and allowing modification of
> // "sliced" Bytes since we can now support semantics similar to Golang's slices.
> func (b *Bytes) Window(start, end int) *Bytes {
> ```

Even with this PR I still don't think it is safe to modify a window, because calls to `Set` could cause a reallocation of `Bytes.buffer`. Go slices don't have this problem (assignments to index expressions will never cause a reallocation of the underlying array).

Random thought: it might be more effective for "Window" to be a separate type that is simply missing the modification methods. Then these checks would happen at compile time instead of at (test) runtime. (But that's not a suggestion for this PR; just a random thought.)



_Code quote:_
```Go
// TODO(yuzefovich): consider renaming it to Slice and allowing modification of
// "sliced" Bytes since we can now support semantics similar to Golang's slices.
```
___
*[pkg/col/coldata/bytes.go, line 304 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxg3qU79tZd7ZNJGXgv:-Mxg3qU79tZd7ZNJGXgw:be8oz5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L304)):*
> ```Go
> 		// we can set the new length accordingly (the elements will be set
> 		// below).
> 		b.zero(newLen)
> ```

Why do we zero these? If we zero them then we won't be able to reuse spots in buffer for non-inlined elements.
___
*[pkg/col/coldata/bytes.go, line 307 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxg4ikT4NZ9LZlsbqQU:-Mxg4ikT4NZ9LZlsbqQV:bcn5h2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L307)):*
> ```Go
> 		b.elements = b.elements[:newLen]
> 	}
> 	for i := 0; i < len(srcElementsToCopy); i++ {
> ```

We might also need the reverse loop case from `CopySlice` when `b == src && destIdx > srcStartIdx`. Come to think of it, seems like `AppendSlice` could simply call `CopySlice` after ensuring the capacity of b but maybe I'm overthinking it.
___
*[pkg/sql/colexec/colexecspan/span_encoder_tmpl.go, line 127 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MxgC4Yj1EiOPPYLrdst:-MxgC4Yj1EiOPPYLrdsu:bi4u0xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/sql/colexec/colexecspan/span_encoder_tmpl.go#L127)):*
> ```Go
> func (op *_OP_STRING) next(batch coldata.Batch, startIdx, endIdx int) *coldata.Bytes {
> 	oldBytesSize := op.outputBytes.Size()
> 	if op.outputBytes == nil || op.outputBytes.Len() < endIdx-startIdx {
> ```

nit: Couldn't this check capacity instead of length? (But then I guess we would need a `Cap` method and a way to set the length?)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I added another cleanup commit into this PR. I could see the last two commits not passing the bar for being merged during the stability period, still personally I'd like to get them in. Curious to hear your thoughts.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-Mxhdau_A8aBfurhDs3K:b-lcd06s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @michae2)

___
*[pkg/col/coldata/bytes.go, line 164 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvQnobQ6XDcpKnieyBr:-MxgW9tx9fN909d7rAdn:b-vt6y0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/25898336c28b768e86c25f779b23455503b3cccc/pkg/col/coldata/bytes.go#L164)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This is fine. No need to squeeze any more byte juice.

nit: But it might be nice to reverse the meaning of the `inlined` boolean so that the zero `element` was a zero-length inlined `[]byte` rather than a zero-length `[]byte` at offset 0 of the buffer. (Maybe you meant to do that, since you wrote `nonInlined` instead of `inlined` in your comment above?)
</blockquote></details>

I did reverse the meaning from `nonInlined` to `inlined` in the latest set of changes (like a couple of days ago).
___
*[pkg/col/coldata/bytes.go, line 186 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxfw3DRCYOJa_5YuD0g:-MxgWc52BGSyaxJYbhwL:b9qb2ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L186)):*
> calls to `Set` could cause a reallocation of `Bytes.buffer`

I think everything would still work, no?

Let's think through an example (imagine we have zero inlinable space):
```
b := NewBytes(2)
b.Set(0, []byte("zero"))
b.Set(1, []byte("one"))
```
At this point we have:
```
buffer = []byte("zeroone")
elements = []element{{offset: 0, len: 4}, {offset: 4, len: 3}}
```
Now we do `b.Window(0, 1).Set(0, "long_string")` which will cause reallocation of the buffer:
```
buffer = []byte("zeroonelong_string")
elements = []element{{offset: 7, len: 11}, {offset: 4, len: 3}}
```
We have successfully overwritten the zeroth element in the window as well as in the original vector. The only thing is that `buffer[:4]` is now garbage and will never be used again, unless the vector is reset.

Am I missing something?

> Random thought: it might be more effective for "Window" to be a separate type that is simply missing the modification methods. Then these checks would happen at compile time instead of at (test) runtime. (But that's not a suggestion for this PR; just a random thought.)

I've thought a bit about this too previously. We could imagine having  something `ImmutableVec` interface that doesn't include any mutation methods and making `Vec` interface embed `ImmutableVec` and add those mutation methods. This will - probably - require quite a bit of changes in the plumbing, and I thought it wasn't worth it. Additionally, it will likely come with some cost since we're performing all operations on the well-typed structs, so if we were to pull methods like `Get` and `Set` into the interface, we'd have to pay the interface dispatch cost. Alternatively, we would need to have separate well-typed structs - something like `ImmutableBytes` and `Bytes`.

OTOH we only have a handful of use cases where we're using these "windows" (like when we're spilling to disk from an in-memory operator in order to "export the buffered data"), and I believe if in those cases we actually did modify the window, then `fakedist-disk` logic test config would have crashed or returned incorrect results.
___
*[pkg/col/coldata/bytes.go, line 304 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxg3qU79tZd7ZNJGXgv:-Mxgag7i2dBtWsHplkuM:bxfpsg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L304)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why do we zero these? If we zero them then we won't be able to reuse spots in buffer for non-inlined elements.
</blockquote></details>

Hm. I think I wanted to maintain an invariant that all `element`s in `b.elements[len:cap]` range are uninitialized which makes `Reset` faster when `len < cap`.

Consider the following scenario with zero inlinable space:
```
b := NewBytes(2)
b.Set(0, "foo")
b.Set(1, "bar")
b.Set(0, "fooo")
// buffer = "foobarfooo"
// elements = {{offset: 6, len: 4}, {offset: 3, len: 3}}

// Truncate b to length 1.
b.AppendSlice(b, 1, 0, 0)
// buffer = "foobarfooo"
// elements = {{offset: 6, len: 4}, {offset: 3, len: 3}}, len(elements) = 1

// Reset with zeroing out elements only up to len(b.elements)
b.Reset()
// buffer: len = 0, cap = 10
// elements = {{}, {offset: 3, len: 3}}, len(elements) = 1

b.Set(0, "zero")
// buffer = "zero", cap = 10
// elements = {{offset: 0, len: 4}, {offset: 3, len: 3}}, len(elements) = 1

b2 := NewBytes(1)
b2.Set(0, "one")
// Here we will get corruption because b.elements[1] thinks it can reuse
// buffer[3:6], but the zeroth element uses buffer[0:4].
b.AppendSlice(b2, 1 /* destIdx */, 0 /* startIdx */, 1 /* endIdx */)
```

If this is right, then we either have to zero out all elements after `newLen` in `AppendSlice` or to zero out all elements up to `cap(b.elements)` in `Reset`.

----

Update: this is, indeed, a problem, and I converted this snippet into a unit test. I also decided to switch to zeroing out elements up to the capacity in `Reset`.
___
*[pkg/col/coldata/bytes.go, line 307 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxg4ikT4NZ9LZlsbqQU:-MxhU0924C0jmkEq8AZ0:bd9mmsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L307)):*
`AppendSlice` from the vector into itself is only allowed when `srcStartIdx == srcEndIdx` in which case there is no overlap, so we don't have to be too smart about it. This is why I opened #76529 some time ago.

> Come to think of it, seems like `AppendSlice` could simply call `CopySlice` after ensuring the capacity of b

That's a nice observation - I factored out a common helper method using the same optimization as we had in the append case. I also added some BCEs, and the benchmarks have improved for `CopySlice`.
___
*[pkg/sql/colexec/colexecspan/span_encoder_tmpl.go, line 127 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MxgC4Yj1EiOPPYLrdst:-Mxhc2O76yOgUwvKB4A4:b-56eov6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/sql/colexec/colexecspan/span_encoder_tmpl.go#L127)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: Couldn't this check capacity instead of length? (But then I guess we would need a `Cap` method and a way to set the length?)
</blockquote></details>

Given that we're now using `Set`s in this file, checking the length is the right thing to do - `Set`s are only allowed in `[0, len]` range.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I think it would be fine to merge the last two commits.

> Reviewed 1 of 15 files at r1, 2 of 8 files at r8, 8 of 8 files at r13, 5 of 5 files at r14, 2 of 2 files at r15, 3 of 3 files at r16, 17 of 17 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MxkM-VW7Z_KPXFbyO4C:bh4rnby)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/col/coldata/bytes.go, line 186 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxfw3DRCYOJa_5YuD0g:-MxjlXOf1e9ZdbEdjfx8:ba6sqls) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L186)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

> calls to `Set` could cause a reallocation of `Bytes.buffer`

I think everything would still work, no?

Let's think through an example (imagine we have zero inlinable space):
```
b := NewBytes(2)
b.Set(0, []byte("zero"))
b.Set(1, []byte("one"))
```
At this point we have:
```
buffer = []byte("zeroone")
elements = []element{{offset: 0, len: 4}, {offset: 4, len: 3}}
```
Now we do `b.Window(0, 1).Set(0, "long_string")` which will cause reallocation of the buffer:
```
buffer = []byte("zeroonelong_string")
elements = []element{{offset: 7, len: 11}, {offset: 4, len: 3}}
```
We have successfully overwritten the zeroth element in the window as well as in the original vector. The only thing is that `buffer[:4]` is now garbage and will never be used again, unless the vector is reset.

Am I missing something?

> Random thought: it might be more effective for "Window" to be a separate type that is simply missing the modification methods. Then these checks would happen at compile time instead of at (test) runtime. (But that's not a suggestion for this PR; just a random thought.)

I've thought a bit about this too previously. We could imagine having  something `ImmutableVec` interface that doesn't include any mutation methods and making `Vec` interface embed `ImmutableVec` and add those mutation methods. This will - probably - require quite a bit of changes in the plumbing, and I thought it wasn't worth it. Additionally, it will likely come with some cost since we're performing all operations on the well-typed structs, so if we were to pull methods like `Get` and `Set` into the interface, we'd have to pay the interface dispatch cost. Alternatively, we would need to have separate well-typed structs - something like `ImmutableBytes` and `Bytes`.

OTOH we only have a handful of use cases where we're using these "windows" (like when we're spilling to disk from an in-memory operator in order to "export the buffered data"), and I believe if in those cases we actually did modify the window, then `fakedist-disk` logic test config would have crashed or returned incorrect results.
</blockquote></details>

Splitting this `b.Window(0, 1).Set(0, "long_string")` out into:

```go
c := b.Window(0, 1)
c.Set(0, "long_string")
```

Won't the call to `c.Set` assign the reallocated buffer to `c.buffer` but not `b.buffer`?

Anyway, I'm not requesting any changes to the code :laughing: just agreeing with the current immutableness of Window.
___
*[pkg/col/coldata/bytes.go, line 304 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxg3qU79tZd7ZNJGXgv:-MxjyBaf30Awb5HZL3p2:bcztznv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L304)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm. I think I wanted to maintain an invariant that all `element`s in `b.elements[len:cap]` range are uninitialized which makes `Reset` faster when `len < cap`.

Consider the following scenario with zero inlinable space:
```
b := NewBytes(2)
b.Set(0, "foo")
b.Set(1, "bar")
b.Set(0, "fooo")
// buffer = "foobarfooo"
// elements = {{offset: 6, len: 4}, {offset: 3, len: 3}}

// Truncate b to length 1.
b.AppendSlice(b, 1, 0, 0)
// buffer = "foobarfooo"
// elements = {{offset: 6, len: 4}, {offset: 3, len: 3}}, len(elements) = 1

// Reset with zeroing out elements only up to len(b.elements)
b.Reset()
// buffer: len = 0, cap = 10
// elements = {{}, {offset: 3, len: 3}}, len(elements) = 1

b.Set(0, "zero")
// buffer = "zero", cap = 10
// elements = {{offset: 0, len: 4}, {offset: 3, len: 3}}, len(elements) = 1

b2 := NewBytes(1)
b2.Set(0, "one")
// Here we will get corruption because b.elements[1] thinks it can reuse
// buffer[3:6], but the zeroth element uses buffer[0:4].
b.AppendSlice(b2, 1 /* destIdx */, 0 /* startIdx */, 1 /* endIdx */)
```

If this is right, then we either have to zero out all elements after `newLen` in `AppendSlice` or to zero out all elements up to `cap(b.elements)` in `Reset`.

----

Update: this is, indeed, a problem, and I converted this snippet into a unit test. I also decided to switch to zeroing out elements up to the capacity in `Reset`.
</blockquote></details>

Interesting! Thank you for the example, I wasn't thinking about `Reset` at all (or using `AppendSlice` for truncation, ha). Zeroing up to capacity in `Reset` makes sense to me.
___
*[pkg/col/coldata/bytes.go, line 137 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxk7bvoDfeDSB9SQqWf:-Mxk7bvoDfeDSB9SQqWg:b63ehz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eac85161c53c6b1adfb1b5ccac0e5a23bb69ced/pkg/col/coldata/bytes.go#L137)):*
> ```Go
> 	// Check if we there was an old non-inlined value we can overwrite.
> 	if !e.inlined && e.header.cap >= len(v) {
> 		copy(e.getNonInlined(b)[:len(v)], v)
> ```

nit: If `element.getNonInlined` took a length argument then we wouldn't need the `[:len(v)]` expression here. But maybe that's too finicky.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the reviews and the feedback!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-:-MxkQMyp9RKoJ-OoXaDY:b-k1dy3g)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @michae2)

___
*[pkg/col/coldata/bytes.go, line 186 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxfw3DRCYOJa_5YuD0g:-MxkOqrR1WTI1GJkDLey:b-v8tc1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b423147c6e44929e852298b7546331cf014f0ec/pkg/col/coldata/bytes.go#L186)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Splitting this `b.Window(0, 1).Set(0, "long_string")` out into:

```go
c := b.Window(0, 1)
c.Set(0, "long_string")
```

Won't the call to `c.Set` assign the reallocated buffer to `c.buffer` but not `b.buffer`?

Anyway, I'm not requesting any changes to the code :laughing: just agreeing with the current immutableness of Window.
</blockquote></details>

Oh, indeed. Removed the TODO.
___
*[pkg/col/coldata/bytes.go, line 137 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-Mxk7bvoDfeDSB9SQqWf:-MxkQ0Hk1QssrWnNHeK8:b-tqohwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eac85161c53c6b1adfb1b5ccac0e5a23bb69ced/pkg/col/coldata/bytes.go#L137)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: If `element.getNonInlined` took a length argument then we wouldn't need the `[:len(v)]` expression here. But maybe that's too finicky.
</blockquote></details>

I refactored the code a bit to remove this.


<!-- Sent from Reviewable.io -->

Azhng commented on pkg/col/coldata/bytes.go:
```diff
@@ -17,161 +17,170 @@ import (
+	// bytes value begins.
+	bufferOffset int
+	len          int
+	cap          int
```
Drive by comment, should we be explicit to use `int64` here? Since the size of `int` is platform dependent. On 32-bit platform, this would mean the `sliceHeader` would be 12 bytes instead of 24 bytes intended.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76202)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/col/coldata/bytes.go, line 77 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/76202#-MvL7GIHMjacqw9NnLxn-r18-77:-MxokhqN7fBO8VvLiYi-:b-b2426v) ([raw file](https://github.com/cockroachdb/cockroach/blob/812ade9f024cb034fae63f9d879a3f41896ab357/pkg/col/coldata/bytes.go#L77)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Drive by comment, should we be explicit to use `int64` here? Since the size of `int` is platform dependent. On 32-bit platform, this would mean the `sliceHeader` would be 12 bytes instead of 24 bytes intended. 
</blockquote></details>

I thought about it and chose to keep using `int`.

Everything will still work correctly on 32-bit systems, we'll just have less inlinable space (18 bytes instead of 30) and we won't be using the whole size of the object in the Golang's [size class](https://go.dev/src/runtime/sizeclasses.go) (`element` will take up 20 bytes, and it'll be in 24 byte object class, so we could have used 4 bytes for more padding; alternatively, we could have used less padding to get into 16 byte class).

I think that 32-bit systems are quite rare nowadays, and we usually just assume that we're running on a 64-bit system.


<!-- Sent from Reviewable.io -->
